### PR TITLE
Remove all minikube dependencies from drivers

### PR DESCRIPTION
**Pull request by sharifelgamal**
All the config not related to networking gets overridden in the config passed in from the client, so there's never a need to set any defaults.

### CODE REVIEW

File: pkg/drivers/hyperkit/driver.go [hunk 1]
```diff
@@ -121,7 +117,7 @@ func (d *Driver) Create() error {
 
 // DriverName returns the name of the driver
 func (d *Driver) DriverName() string {
-	return constants.DriverHyperkit
+	return "hyperkit"
```
Comment (josedonizetti) [inline pkg/drivers/hyperkit/driver.go]: can we make it a constant inside the driver?

File: pkg/drivers/kvm/kvm.go [hunk 2]
```diff
@@ -224,7 +214,7 @@ func (d *Driver) GetSSHHostname() (string, error) {
 
 // DriverName returns the name of the driver
 func (d *Driver) DriverName() string {
-	return constants.DriverKvm2
+	return "kvm2"
```
Comment (josedonizetti) [inline pkg/drivers/kvm/kvm.go]: can we make it a constant inside the driver?

File: pkg/drivers/hyperkit/driver.go [hunk 3]
```diff
@@ -268,13 +264,22 @@ func (d *Driver) Start() error {
...
+	}
+
+	// Implement a retry loop without calling any minikube code
+	for i := 0; i < 30; i++ {
+		err = getIP()
+		if err == nil {
+			break
 		}
-		return nil
+		if err.Error() != fmt.Sprintf(IPErrorMessage, mac) {
```
Comment (josedonizetti) [inline pkg/drivers/hyperkit/driver.go]: hey, do you think it's worth having a custom error and testing the type? comparing error strings seems ugly :/

File: pkg/drivers/kvm/kvm.go (hunk 2 above)
Comment (sharifelgamal) [reply pkg/drivers/kvm/kvm.go]: yep, done.

File: pkg/drivers/hyperkit/driver.go (hunk 1 above)
Comment (sharifelgamal) [reply pkg/drivers/hyperkit/driver.go]: yep, done.

File: pkg/drivers/hyperkit/driver.go (hunk 3 above)
Comment (sharifelgamal) [reply pkg/drivers/hyperkit/driver.go]: yeah I wasn't wild about text parsing either, and the custom error type is easy to add. done.

File: pkg/drivers/hyperkit/driver.go [hunk 4]
```diff
@@ -52,6 +49,7 @@ const (
 	permErr         = "%s needs to run with elevated permissions. " +
 		"Please run the following command, then try again: " +
 		"sudo chown root:wheel %s && sudo chmod u+s %s"
+	driverHyperkit = "hyperkit"
```
Comment (afbjorklund) [inline pkg/drivers/hyperkit/driver.go]: or just use the string?

File: pkg/drivers/kvm/kvm.go [hunk 5]
```diff
@@ -91,6 +87,7 @@ const (
 	qemusystem                = "qemu:///system"
 	defaultPrivateNetworkName = "minikube-net"
 	defaultNetworkName        = "default"
+	driverKvm2 = "kvm2"
```
Comment (afbjorklund) [inline pkg/drivers/kvm/kvm.go]: or just use the string ?

File: pkg/drivers/kvm/kvm.go [hunk 6]
```diff
@@ -224,7 +215,7 @@ func (d *Driver) GetSSHHostname() (string, error) {
 
 // DriverName returns the name of the driver
 func (d *Driver) DriverName() string {
-	return constants.DriverKvm2
+	return driverKvm2
```
Comment (afbjorklund) [inline pkg/drivers/kvm/kvm.go]: "kvm2"

File: pkg/drivers/hyperkit/driver.go [hunk 7]
```diff
@@ -121,7 +118,7 @@ func (d *Driver) Create() error {
 
 // DriverName returns the name of the driver
 func (d *Driver) DriverName() string {
-	return constants.DriverHyperkit
+	return driverHyperkit
```
Comment (afbjorklund) [inline pkg/drivers/hyperkit/driver.go]: "hyperkit"

File: pkg/drivers/hyperkit/network.go [hunk 8]
```diff
@@ -78,7 +80,7 @@ func getIPAddressFromFile(mac, path string) (string, error) {
 			return dhcpEntry.IPAddress, nil
 		}
 	}
-	return "", fmt.Errorf("could not find an IP address for %s", mac)
+	return "", fmt.Errorf(IPErrorMessage, mac)
```
Comment (afbjorklund) [inline pkg/drivers/hyperkit/network.go]: would prefer having the printf inline here

File: pkg/drivers/hyperkit/driver.go (hunk 1 above)
Comment (afbjorklund) [reply pkg/drivers/hyperkit/driver.go]: I disagree :-P

File: pkg/drivers/kvm/kvm.go (hunk 2 above)
Comment (afbjorklund) [reply pkg/drivers/kvm/kvm.go]: But why

**Comment by afbjorklund** [comment]
> All the config not related to networking gets overridden in the config passed in from the client, so there's never a need to set any defaults.

I find this strange, since normally it does have some defaults. But I guess we will see, once it has been tried stand-alone (with `docker-machine`)

Then again, those defaults are usually 1 CPU and 1G RAM since that is what Docker requires.

It is Kubernetes that is increasing that to 2 CPU and 2G RAM, or maybe I should kubeadm does.

File: pkg/drivers/kvm/kvm.go (hunk 2 above)
Comment (josedonizetti) [reply pkg/drivers/kvm/kvm.go]: @afbjorklund As those are the driver names I consider they good for a constant. To always have only one place to change it, even though now we don't have many. But as always, no strong feelings. and probably a too simple thing to start a big discussion, so all good to revert it to string only :)

Comment (afbjorklund) [reply pkg/drivers/kvm/kvm.go]: Well, at least it's not using the constants module. But the string is only used in 1 or 2 places ?

It's the not the most important thing, just thought it was an unnecessary step of indirection.

**Comment by afbjorklund** [comment]
This addresses the dependencies found in #4909, but seems like it doesn't build just yet.

```
# k8s.io/minikube/pkg/drivers/hyperkit
pkg/drivers/hyperkit/driver.go:269:12: undefined: TempError
pkg/drivers/hyperkit/driver.go:280:18: impossible type assertion:
	*tempError does not implement error (missing Error method)
```

**Comment by sharifelgamal** [comment]
@minikube-bot OK to test

**Comment by sharifelgamal** [comment]
This should be ready now.

**Comment by afbjorklund** [comment]
Why revert the libmachine version ?

Moving back to the 0.16.1 version re-introduces the AMD bug (yet again)

**Comment by sharifelgamal** [comment]
It was causing issues with go modules, and specifically with the go proxy: 
`curl https://proxy.golang.org/github.com/docker/machine/@v/v0.16.1-0.20190718054102-a555e4f7a8f5.info
not found: github.com/docker/machine@v0.16.1-0.20190718054102-a555e4f7a8f5: invalid pseudo-version: revision a555e4f7a8f5 is not a descendent of preceding tag (v0.16.0)`

I didn't realize that reverting back to the release tag reintroduces a bug, I'll looking into pinning this correctly.

**Comment by afbjorklund** [comment]
> It was causing issues with go modules, and specifically with the go proxy:
No worries, it was my fault for hacking the version. I'll go back to the weirdo `v0.0.0`

**Comment by afbjorklund** [comment]
@sharifelgamal : why aren't we messing with go proxy (and Travis) in a separate PR ?

**Comment by sharifelgamal** [comment]
You're right, it should be a separate PR. The travis issues were causing tests to fail in this PR so I was testing solutions. I'll open a new PR.

**Comment by afbjorklund** [comment]
I can fix the go.mod, in order to make go proxy happy (have never tried go proxy)

**Comment by sharifelgamal** [comment]
#5018 is the new go proxy PR

**Pull request merged**